File: coin_toss.go
# coin_toss.go
/**
 * 🪙 Coin Toss – Advanced Flip Series Analyzer (Go Edition)
 * Features: biased probability, streaks, ASCII chart, export, colorful CLI
 */

package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"math/rand"
	"os"
	"strconv"
	"strings"
	"time"
)

// ─── Colors ──────────────────────────────────────────────────────────────────

const (
	reset  = "\x1b[0m"
	bright = "\x1b[1m"
	dim    = "\x1b[2m"
	red    = "\x1b[31m"
	green  = "\x1b[32m"
	yellow = "\x1b[33m"
	blue   = "\x1b[34m"
	magenta = "\x1b[35m"
	cyan   = "\x1b[36m"
)

func c(str, color string) string {
	return color + str + reset
}

// ─── Core Logic ──────────────────────────────────────────────────────────────

type CoinToss struct {
	prob float64
}

func NewCoinToss(prob float64) *CoinToss {
	return &CoinToss{prob: prob}
}

func (ct *CoinToss) Flip() string {
	if rand.Float64() < ct.prob {
		return "H"
	}
	return "T"
}

func (ct *CoinToss) FlipSeries(n int) []string {
	series := make([]string, n)
	for i := 0; i < n; i++ {
		series[i] = ct.Flip()
	}
	return series
}

// ─── Statistics ─────────────────────────────────────────────────────────────

type Stats struct {
	Heads               int     `json:"heads"`
	Tails               int     `json:"tails"`
	Total               int     `json:"total"`
	HeadsPercent        float64 `json:"headsPercent"`
	TailsPercent        float64 `json:"tailsPercent"`
	LongestRunHeads     int     `json:"longestRunHeads"`
	LongestRunTails     int     `json:"longestRunTails"`
	LongestStreakHeads  int     `json:"longestStreakHeads"`
	LongestStreakTails  int     `json:"longestStreakTails"`
	Runs                int     `json:"runs"`
}

func analyzeSeries(series []string) Stats {
	if len(series) == 0 {
		return Stats{}
	}
	heads := 0
	for _, s := range series {
		if s == "H" {
			heads++
		}
	}
	total := len(series)
	tails := total - heads
	longestRunHeads, longestRunTails := 0, 0
	currentRun := 1
	for i := 1; i < len(series); i++ {
		if series[i] == series[i-1] {
			currentRun++
		} else {
			if series[i-1] == "H" {
				if currentRun > longestRunHeads {
					longestRunHeads = currentRun
				}
			} else {
				if currentRun > longestRunTails {
					longestRunTails = currentRun
				}
			}
			currentRun = 1
		}
	}
	if len(series) > 0 {
		if series[len(series)-1] == "H" {
			if currentRun > longestRunHeads {
				longestRunHeads = currentRun
			}
		} else {
			if currentRun > longestRunTails {
				longestRunTails = currentRun
			}
		}
	}
	runs := 1
	for i := 1; i < len(series); i++ {
		if series[i] != series[i-1] {
			runs++
		}
	}
	return Stats{
		Heads:              heads,
		Tails:              tails,
		Total:              total,
		HeadsPercent:       float64(heads) / float64(total) * 100,
		TailsPercent:       float64(tails) / float64(total) * 100,
		LongestRunHeads:    longestRunHeads,
		LongestRunTails:    longestRunTails,
		LongestStreakHeads: longestRunHeads,
		LongestStreakTails: longestRunTails,
		Runs:               runs,
	}
}

func drawAsciiChart(stats Stats, width int) string {
	if width <= 0 {
		width = 40
	}
	h := stats.HeadsPercent
	t := stats.TailsPercent
	hBar := int(h / 100 * float64(width))
	tBar := int(t / 100 * float64(width))
	line := fmt.Sprintf("H %s%s %.1f%%\n",
		c(strings.Repeat("█", hBar), green),
		strings.Repeat(" ", width-hBar),
		h)
	line += fmt.Sprintf("T %s%s %.1f%%",
		c(strings.Repeat("█", tBar), red),
		strings.Repeat(" ", width-tBar),
		t)
	return line
}

// ─── Export ──────────────────────────────────────────────────────────────────

type ExportData struct {
	Flips     []string `json:"flips"`
	Stats     Stats    `json:"stats"`
	Timestamp string   `json:"timestamp"`
}

func exportResults(series []string, stats Stats, filename string) {
	if filename == "" {
		ts := time.Now().Format("20060102_150405")
		filename = fmt.Sprintf("coin_toss_%s.json", ts)
	}
	data := ExportData{
		Flips:     series,
		Stats:     stats,
		Timestamp: time.Now().Format(time.RFC3339),
	}
	raw, _ := json.MarshalIndent(data, "", "  ")
	os.WriteFile(filename, raw, 0644)
	fmt.Printf("%s\n", c(fmt.Sprintf("✅ Results exported to %s", filename), green))
}

// ─── Main App ──────────────────────────────────────────────────────────────

type CoinApp struct {
	reader      *bufio.Reader
	prob        float64
	toss        *CoinToss
	lastSeries  []string
	lastStats   Stats
}

func NewCoinApp() *CoinApp {
	rand.Seed(time.Now().UnixNano())
	return &CoinApp{
		reader: bufio.NewReader(os.Stdin),
		prob:   0.5,
		toss:   NewCoinToss(0.5),
	}
}

func (app *CoinApp) ask(prompt string) string {
	fmt.Print(prompt)
	line, _ := app.reader.ReadString('\n')
	return strings.TrimSpace(line)
}

func (app *CoinApp) askInt(prompt string, def int) int {
	for {
		ans := app.ask(prompt)
		if ans == "" {
			return def
		}
		if n, err := strconv.Atoi(ans); err == nil {
			return n
		}
		fmt.Println(c("Please enter a number.", yellow))
	}
}

func (app *CoinApp) askFloat(prompt string, def float64) float64 {
	for {
		ans := app.ask(prompt)
		if ans == "" {
			return def
		}
		if n, err := strconv.ParseFloat(ans, 64); err == nil {
			return n
		}
		fmt.Println(c("Please enter a number.", yellow))
	}
}

func (app *CoinApp) showMenu() {
	fmt.Println("\n" + c(strings.Repeat("═", 50), cyan))
	fmt.Println(c("🪙 COIN TOSS ANALYZER", bright+cyan))
	fmt.Println(c(strings.Repeat("═", 50), cyan))
	fmt.Printf("  Probability of Heads: %.2f\n", app.prob)
	fmt.Println(c(strings.Repeat("═", 50), cyan))
	fmt.Println("  1. 🪙 Single Flip")
	fmt.Println("  2. 📊 Series of Flips")
	fmt.Println("  3. 📈 Show Statistics")
	fmt.Println("  4. 🎯 Set Probability")
	fmt.Println("  5. 💾 Export Results")
	fmt.Println("  6. 🔁 Continuous Flip Mode")
	fmt.Println("  0. 🚪 Exit")
	fmt.Println(c(strings.Repeat("═", 50), cyan))
}

func (app *CoinApp) singleFlip() {
	res := app.toss.Flip()
	color := green
	if res == "T" {
		color = red
	}
	fmt.Printf("\n🪙 Result: %s (Heads=%.2f)\n", c(res, color), app.prob)
	app.lastSeries = append(app.lastSeries, res)
	app.lastStats = analyzeSeries(app.lastSeries)
}

func (app *CoinApp) seriesFlips() {
	n := app.askInt("Number of flips (default 10): ", 10)
	if n <= 0 {
		fmt.Println(c("Number must be positive.", red))
		return
	}
	fmt.Printf(c("Flipping %d times...\n", dim), n)
	series := app.toss.FlipSeries(n)
	app.lastSeries = series
	app.lastStats = analyzeSeries(series)
	app.showStats()
}

func (app *CoinApp) showStats() {
	if app.lastStats.Total == 0 {
		fmt.Println(c("No data yet. Do some flips!", yellow))
		return
	}
	s := app.lastStats
	fmt.Println("\n📊 STATISTICS")
	fmt.Println(c(strings.Repeat("─", 30), dim))
	fmt.Printf("  Total Flips: %d\n", s.Total)
	fmt.Printf("  Heads: %d (%.1f%%)\n", s.Heads, s.HeadsPercent)
	fmt.Printf("  Tails: %d (%.1f%%)\n", s.Tails, s.TailsPercent)
	fmt.Printf("  Longest Run (H): %d\n", s.LongestRunHeads)
	fmt.Printf("  Longest Run (T): %d\n", s.LongestRunTails)
	fmt.Printf("  Number of Runs: %d\n", s.Runs)
	fmt.Println("\n📊 Distribution:")
	fmt.Println(drawAsciiChart(s, 40))
}

func (app *CoinApp) setProbability() {
	p := app.askFloat("Probability of Heads (0.0–1.0): ", 0.5)
	if p >= 0 && p <= 1 {
		app.prob = p
		app.toss = NewCoinToss(p)
		fmt.Printf("%s\n", c(fmt.Sprintf("✅ Probability set to %.2f", p), green))
	} else {
		fmt.Println(c("Invalid probability.", red))
	}
}

func (app *CoinApp) exportResults() {
	if app.lastStats.Total == 0 {
		fmt.Println(c("No data to export.", yellow))
		return
	}
	filename := app.ask("Filename (leave empty for auto): ")
	exportResults(app.lastSeries, app.lastStats, filename)
}

func (app *CoinApp) continuousFlip() {
	fmt.Printf("%s\n", c("\n🔁 Continuous Flip Mode (press Ctrl+C to stop)", cyan))
	count := 0
	heads := 0
	tails := 0
	series := []string{}
	for {
		res := app.toss.Flip()
		series = append(series, res)
		if res == "H" {
			heads++
		} else {
			tails++
		}
		count++
		fmt.Printf("\r %d flips: H=%d (%.1f%%) T=%d (%.1f%%)", count, heads, float64(heads)/float64(count)*100, tails, float64(tails)/float64(count)*100)
	}
}

func (app *CoinApp) run() {
	fmt.Print("\033[H\033[2J")
	fmt.Printf("%s\n", c("\n🪙 Coin Toss – Advanced Flip Analyzer", bright+cyan))
	fmt.Printf("%s\n", c("Flip, analyze, and master probability!", dim))

	for {
		app.showMenu()
		choice := app.ask("Your choice: ")
		switch choice {
		case "1":
			app.singleFlip()
		case "2":
			app.seriesFlips()
		case "3":
			app.showStats()
		case "4":
			app.setProbability()
		case "5":
			app.exportResults()
		case "6":
			app.continuousFlip()
			// After continuous mode, we won't return due to infinite loop, but we can handle exit
		case "0":
			fmt.Printf("%s\n", c("👋 Goodbye!", cyan))
			return
		default:
			fmt.Println(c("❌ Invalid choice.", red))
		}
		if choice != "0" && choice != "6" {
			fmt.Print("\nPress Enter to continue...")
			app.reader.ReadString('\n')
		}
	}
}

func main() {
	app := NewCoinApp()
	app.run()
}

File: coin_toss.js
# coin_toss.js
/**
 * 🪙 Coin Toss – Advanced Flip Series Analyzer (Node.js Edition)
 * Features: biased probability, streaks, ASCII chart, export, colorful CLI
 */

const fs = require('fs');
const path = require('path');
const os = require('os');
const readline = require('readline');

// ─── Colors ──────────────────────────────────────────────────────────────────

const colors = {
  reset: '\x1b[0m',
  bright: '\x1b[1m',
  dim: '\x1b[2m',
  red: '\x1b[31m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  magenta: '\x1b[35m',
  cyan: '\x1b[36m',
};

const c = (str, color) => `${color}${str}${colors.reset}`;

// ─── Core Logic ──────────────────────────────────────────────────────────────

class CoinToss {
  constructor(prob = 0.5) {
    this.prob = prob;
  }
  flip() {
    return Math.random() < this.prob ? 'H' : 'T';
  }
  flipSeries(n) {
    const arr = [];
    for (let i = 0; i < n; i++) arr.push(this.flip());
    return arr;
  }
}

// ─── Statistics ─────────────────────────────────────────────────────────────

function analyzeSeries(series) {
  if (!series || series.length === 0) return {};
  const heads = series.filter(s => s === 'H').length;
  const total = series.length;
  const tails = total - heads;
  let longestRunHeads = 0, longestRunTails = 0;
  let currentRun = 1;
  for (let i = 1; i < series.length; i++) {
    if (series[i] === series[i-1]) currentRun++;
    else {
      if (series[i-1] === 'H') longestRunHeads = Math.max(longestRunHeads, currentRun);
      else longestRunTails = Math.max(longestRunTails, currentRun);
      currentRun = 1;
    }
  }
  if (series.length > 0) {
    if (series[series.length-1] === 'H') longestRunHeads = Math.max(longestRunHeads, currentRun);
    else longestRunTails = Math.max(longestRunTails, currentRun);
  }
  let runs = 1;
  for (let i = 1; i < series.length; i++) {
    if (series[i] !== series[i-1]) runs++;
  }
  return {
    heads,
    tails,
    total,
    headsPercent: total ? heads / total * 100 : 0,
    tailsPercent: total ? tails / total * 100 : 0,
    longestRunHeads,
    longestRunTails,
    longestStreakHeads: longestRunHeads,
    longestStreakTails: longestRunTails,
    runs
  };
}

function drawAsciiChart(stats, width = 40) {
  const h = stats.headsPercent || 0;
  const t = stats.tailsPercent || 0;
  const hBar = Math.floor(h / 100 * width);
  const tBar = Math.floor(t / 100 * width);
  let line = `H ${c('█'.repeat(hBar), colors.green)}${' '.repeat(width - hBar)} ${h.toFixed(1)}%\n`;
  line += `T ${c('█'.repeat(tBar), colors.red)}${' '.repeat(width - tBar)} ${t.toFixed(1)}%`;
  return line;
}

// ─── Export ──────────────────────────────────────────────────────────────────

function exportResults(series, stats, filename = null) {
  if (!filename) {
    const ts = new Date().toISOString().replace(/[:.]/g, '').slice(0,14);
    filename = `coin_toss_${ts}.json`;
  }
  const data = {
    flips: series,
    stats,
    timestamp: new Date().toISOString()
  };
  fs.writeFileSync(filename, JSON.stringify(data, null, 2));
  console.log(c(`✅ Results exported to ${filename}`, colors.green));
}

// ─── Main App ──────────────────────────────────────────────────────────────

class CoinApp {
  constructor() {
    this.rl = readline.createInterface({ input: process.stdin, output: process.stdout });
    this.prob = 0.5;
    this.toss = new CoinToss(this.prob);
    this.lastSeries = [];
    this.lastStats = {};
  }

  _ask(prompt) {
    return new Promise(resolve => this.rl.question(prompt, resolve));
  }

  _askInt(prompt, def) {
    return this._ask(prompt).then(ans => {
      const n = parseInt(ans.trim());
      return isNaN(n) ? def : n;
    });
  }

  _askFloat(prompt, def) {
    return this._ask(prompt).then(ans => {
      const n = parseFloat(ans.trim());
      return isNaN(n) ? def : n;
    });
  }

  async showMenu() {
    console.log('\n' + c('═'.repeat(50), colors.cyan));
    console.log(c('🪙 COIN TOSS ANALYZER', colors.bright + colors.cyan));
    console.log(c('═'.repeat(50), colors.cyan));
    console.log(`  Probability of Heads: ${this.prob.toFixed(2)}`);
    console.log(c('═'.repeat(50), colors.cyan));
    console.log('  1. 🪙 Single Flip');
    console.log('  2. 📊 Series of Flips');
    console.log('  3. 📈 Show Statistics');
    console.log('  4. 🎯 Set Probability');
    console.log('  5. 💾 Export Results');
    console.log('  6. 🔁 Continuous Flip Mode');
    console.log('  0. 🚪 Exit');
    console.log(c('═'.repeat(50), colors.cyan));
  }

  async singleFlip() {
    const res = this.toss.flip();
    const color = res === 'H' ? colors.green : colors.red;
    console.log(`\n🪙 Result: ${c(res, color)} (Heads=${this.prob.toFixed(2)})`);
    this.lastSeries.push(res);
    this.lastStats = analyzeSeries(this.lastSeries);
  }

  async seriesFlips() {
    const n = await this._askInt('Number of flips (default 10): ', 10);
    if (n <= 0) { console.log(c('Number must be positive.', colors.red)); return; }
    console.log(c(`Flipping ${n} times...`, colors.dim));
    const series = this.toss.flipSeries(n);
    this.lastSeries = series;
    this.lastStats = analyzeSeries(series);
    this.showStats();
  }

  showStats() {
    if (!this.lastStats || !this.lastStats.total) {
      console.log(c('No data yet. Do some flips!', colors.yellow));
      return;
    }
    const s = this.lastStats;
    console.log('\n📊 STATISTICS');
    console.log(c('─'.repeat(30), colors.dim));
    console.log(`  Total Flips: ${s.total}`);
    console.log(`  Heads: ${s.heads} (${s.headsPercent.toFixed(1)}%)`);
    console.log(`  Tails: ${s.tails} (${s.tailsPercent.toFixed(1)}%)`);
    console.log(`  Longest Run (H): ${s.longestRunHeads}`);
    console.log(`  Longest Run (T): ${s.longestRunTails}`);
    console.log(`  Number of Runs: ${s.runs}`);
    console.log('\n📊 Distribution:');
    console.log(drawAsciiChart(s));
  }

  async setProbability() {
    const p = await this._askFloat('Probability of Heads (0.0–1.0): ', 0.5);
    if (p >= 0 && p <= 1) {
      this.prob = p;
      this.toss = new CoinToss(this.prob);
      console.log(c(`✅ Probability set to ${p.toFixed(2)}`, colors.green));
    } else {
      console.log(c('Invalid probability.', colors.red));
    }
  }

  async exportResults() {
    if (!this.lastStats || !this.lastStats.total) {
      console.log(c('No data to export.', colors.yellow));
      return;
    }
    const filename = await this._ask('Filename (leave empty for auto): ');
    exportResults(this.lastSeries, this.lastStats, filename.trim() || null);
  }

  async continuousFlip() {
    console.log(c('\n🔁 Continuous Flip Mode (press Ctrl+C to stop)', colors.cyan));
    let count = 0, heads = 0, tails = 0;
    const series = [];
    const handler = () => {
      console.log('\n');
      this.lastSeries = series;
      this.lastStats = analyzeSeries(series);
      this.showStats();
      process.exit(0);
    };
    process.on('SIGINT', handler);
    while (true) {
      const res = this.toss.flip();
      series.push(res);
      if (res === 'H') heads++; else tails++;
      count++;
      process.stdout.write(`\r ${count} flips: H=${heads} (${(heads/count*100).toFixed(1)}%) T=${tails} (${(tails/count*100).toFixed(1)}%)`);
    }
    // The loop never exits; SIGINT triggers handler
  }

  async run() {
    console.clear();
    console.log(c('\n🪙 Coin Toss – Advanced Flip Analyzer', colors.bright + colors.cyan));
    console.log(c('Flip, analyze, and master probability!', colors.dim));

    while (true) {
      await this.showMenu();
      const choice = await this._ask('Your choice: ');
      switch (choice.trim()) {
        case '1': await this.singleFlip(); break;
        case '2': await this.seriesFlips(); break;
        case '3': this.showStats(); break;
        case '4': await this.setProbability(); break;
        case '5': await this.exportResults(); break;
        case '6': await this.continuousFlip(); break;
        case '0':
          console.log(c('👋 Goodbye!', colors.cyan));
          this.rl.close();
          return;
        default:
          console.log(c('❌ Invalid choice.', colors.red));
      }
      if (choice !== '0' && choice !== '6') {
        console.log('\nPress Enter to continue...');
        await this._ask('');
      }
    }
  }
}

// ─── Main ────────────────────────────────────────────────────────────────────

const main = async () => {
  try {
    const app = new CoinApp();
    await app.run();
  } catch (e) {
    console.error(c(`❌ Unexpected error: ${e.message}`, colors.red));
    process.exit(1);
  }
};

main();
